track2/src/icsi-scenic-tools-20120105/dpwelib-1.01/sndcat.man
.de Sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.               \" Path Name
.               \"      uses Courier fonts for V4.0
.de PN
.ie t \&\f(CB\\$1\f(NR\\$2
.el \fI\\$1\fP\\$2
..
.		\" The following macros added by dpwe for Ultrix 4..
.               \" Manual section reference
.               \"      uses Courier fonts for V4.0
.de MS
.ie t \&\f(CB\\$1\f(NR(\\$2)\\$3
.el \fI\\$1\fP(\\$2)\\$3
..
.de EX		\" Begin Example
.ie \\n(.$ .nr EX \\$1n
.el \{\
.	ie n .nr EX 0n
.	el .nr EX 0n
.\}
.in +\\n(EXu
.if n .sp 1
.if t .sp .5v
.nf
.CW
.ft CB
..
.de EE		\" End example
.in -\\n(EXu
.fi
.}f             \" restore basic text attributes
.if n .sp 1
.if t .sp .5v
..
.TH SNDCAT 1 "$Date: 2012/01/05 21:21:40 $" dpwe   \" $Header: /u/drspeech/repos/dpwelib/sndcat.man,v 1.8 2012/01/05 21:21:40 dpwe Exp $
.SH NAME
sndcat - Copy between soundfiles using the snd interface library
.SH SYNOPSIS
\fBsndcat
[\fIoptions\fB]
[-o \fIoutputSoundfile\fB]
[\fIinputSoundFiles ...\fB]
\fR
.SH DESCRIPTION
.PN sndcat
Copies standard input, or any number of input soundfiles specified on the 
command line, to an output soundfile, accessing all the files through the 
.MS snd 3 
library, thereby permitting automatic selection of soundfile format, 
access to compressed streams, etc.  As such, it encompasses some of 
the functions of the existing 
.MS sndrfmt 1 ,
.MS sndplay 1 , 
and 
.MS sndrecord 1 
programs.
.SH OPTIONS
Options are shown as alternates (separated by "|") with minimal 
abbreviations (the part before the "?"), according to the conventions 
of 
.MS cle 3 .
Defaults are shown parens.
.TP 12
.B \-o?utputfilename \fIname\fR ("-")
Name of the output file to be written.  If not specified (or if it is 
"-"), output will 
be written to standard output.  However, if -query is set, no output 
is written at all.
.TP 12
.B \-S|\-ip?format \fIkey\fR (none)
Input sound file format, defined by one of the 
.MS sndf 3 
tokens.  If none is specified, the program will attempt to guess 
the soundfile format from its first few bytes.
.TP 12
.B \-T|\-op?format \fIkey\fR (none)
As ipformat, but for output soundfile.  If none is specified, it 
will copy the format of the input file.
.TP 12
.B \-k|\-sk?ip \fIskiptime \fR(0.0)
Skip over the first part of sound, skipping this much time's worth in 
seconds.  As with -d and -e, this option uses
.MS cle 3 's
CLE_TIME type, so times in the form [[hh:]mm:]ss[.sss] can be used.
.TP 12
.B \-K|\-skipf?rames \fIskipFrames\fR (0)
Skip over this many frames at the beginning of the sound (after reading the 
header).  If a skip time is specified as well, skipframes takes precedence.
.TP 12
.B \-d?uration \fIdurtime \fR(0.0)
If greater than zero, force the output sound to be exactly this long.  
If zero, the input sound will be copied to EOF.
.TP 12
.B \-D|\-durf?rames \fIdurFrames\fR (0)
If greater than zero, force the output sound to have exactly this many frames. 
Takes precedence over -duration.
.TP 12
.B \-e?ndtime \fIendtime \fR(0.0)
If greater than skiptime, force the output sound to extend just to 
this point in the original sound (extending it if necessary).  This is 
an alternative to -duration, -d and -D take precedence over -e and -E.  If it 
is less than or equal to starttime, it is ignored.
.TP 12
.B \-E|\-endf?rame \fIendFrames\fR (0)
If greater than effective skipframes, and if duration has not been 
specified by other means (-d, -D or -e), only frames prior to this one 
(counted from the start of the input file) will be transferred.  
Specifying in frames takes precedence (once again) over specifying in 
seconds.
.TP 12
.B \-c?hans|\-chann?els \fIchans\fR (0)
If greater than zero, force the output sound to have this many channels.  
The distribution between input and output channels is controlled by 
\f4-chanmode\f1.
.TP 12
.B \-chanm?ode|\-cm?ode \fImode\fR ("mono")
Control how sound is distributed when input and output have different 
numbers of channels.  In the default mode, when copying a single channel 
file to a multi-channel file, each output channel is a copy of the input 
channel; when copying a multi-channel file to a single-channel output, the 
input channels are all averaged together. Alternatively, you might wish 
to extract a single channel from a multichannel input sound, or to write 
to just one channel of a multichannel output sound.  In this 
case, the chanmode should be set to "A" thru "H" for selecting the first 
thru eighth channels respectively.  "0" to "7" are aliases, and "L" and "R" 
are aliases for "A" and "B".  Lower case is equivalent.
.TP 12
.B \-f?ormat|-fmt \fIformat\fR (none)
A sample coding can be specified for the output, which otherwise will 
attempt to copy the input. Acceptable tokens are:
.P
.ta .5i +1.5i
.nf
	\f4a?law\f1	a-law bytes (8 bit/sample)
	\f4u?law|mu?law\f1	u-law bytes (8 bit/sample)
	\f4c?har\f1	linear bytes (8 bit/sample)
	\f4o?ffsetchar\f1	linear bytes offset so that zero is at 0x80
	\f4s?short|i16\f1	linear short ints (16 bits/sample)
	\f4l?long|i32\f1	linear long ints (32 bit/sample)
	\f4f?loat|f32\f1	floating point (32 bit/sample)
	\f4d?ouble|f64\f1	double precision floats (64 bit/sample)
.fi
.TP 12
.B \-r?ate|\-sr?ate|\-s?amplerate \fIsampleRate\fR
Force the output soundfile to be tagged with the specified sampling rate 
(without resampling).
.TP 12
.B \-g?ain \fIgain\fR (1.0)
Scale the sound data by the specified factor.
.TP 12
.B \-n?omerge\fR (no)
The default action of sndcat (since the introduction of this option 
anyway) is to concatenate all soundfiles into a single output sound. 
With this option, each input sound will be separated by a new 
soundfile header in the output sound, suitable for reading with 
-multi later on.
.TP 12
.B \-mu?lti\fR (no)
If specified, the program will check to see if it can read multiple 
sounds from each input file or stream; this is disabled by default to 
avoid forcing the program to seek to the end of possibly very long 
compressed streams before closing each file.
.TP 12
.B \-q?uery
Query the input sound files only; do not write any output
.TP 12
.B \-v?erbose
Verbose mode - print information on each input soundfile.
.SH EXAMPLES
.I sndcat
uses the 
.MS snd 3
library to access different file formats.  This allows it to read and 
write MP3 format files, provided the appropriate encoder/decoder 
are installed (currently mpg123 and lame).  However, this interface 
is clumsy; MP3 encode/decode is based on matching the .mp3 extension 
in the file name, and employs underlying raw audio streams.  Thus, 
the input file parameters must be explicitly specified by the 
PCM format parameters.  Thus, to extract the fourth minute from 
a stereo 44kHz mp3 file to another, mono-mixed MP3 file:
.EX
sndcat -k 3:00 -d 1:00 infile.mp3 -o out.mp3 -S PCM/R44C2 -T PCM -c 1 
.EE
Support for flac files is similar, based on the external flac binary.
.SH BUGS
I'm not sure why I need the -multi option, or what I've broken by 
introducing it.  Without it, extracting one small section led to 
decompressing the whole file.
.PP
.I sndcat 
will allow you to merge several soundfiles together.  But if their 
formats are not compatible, you may not get what you want.  In 
particular, if the input soundfiles have different sampling rates 
they will not be resampled, but will be copied to a single output 
file whose sampling rate is taken from the first file (unless -nomerge).
.SH NOTES
This program was created to allow access to the "shorten" interface in 
.MS snd 3 .
It functions as an equivalent to 
.MS sndrfmt 1 , 
but accessing files through the more general
.MS snd 3 
layer, rather than 
.MS sndf 3
as that program does.
.SH SEE ALSO
.MS snd 3
.SH AUTHOR
Dan Ellis, 
.B dpwe@icsi.berkeley.edu


track2/src/icsi-scenic-tools-20120105/dpwelib-1.01/sndcat.c
/*
 * sndcat.c
 *	
 * Copy between soundfiles using the snd.h library.
 *
 * 1998jun02 dpwe@icsi.berkeley.edu
 * $Header: /u/drspeech/repos/dpwelib/sndcat.c,v 1.18 2011/03/09 02:06:06 davidj Exp $
 */

#include <dpwelib.h>	/* std includes, defs like stdio.h */
#include <stdlib.h>
#include <snd.h>	/* soundfile handling functions */
#include <cle.h>


char *outputfilename = NULL;

int verb = FALSE;	/* verbosity flag */
int query = FALSE;	/* just querying flag */

/*int sfifmt = 0;		/* input soundfile format requested */
/*int sfofmt = 0;		/* output soundfile format requested */
char *sfifmtname = NULL;
char *sfofmtname = NULL;
float skiptime = 0.0;	/* seconds to skip */
float durtime = 0.0;	/* duration in seconds */
float endtime = 0.0;	/* end time in seconds */
int skipframes = 0;	/* frames to skip */
int durframes = 0;	/* total frames */
int endframe = 0;	/* absolute end frame */
int chans = 0;		/* output chans */
int chanmode = SCMD_MONO;	/* should channel reduction mix or extract? */
float srate = 0.0;	/* output srate */
int sampfmt = 0;	/* output sample format */
float gain = 1.0;	/* gain */
int help = 0;

int domulti = 0;	/* attempt to read multiple sounds in stream */
int nomerge = 0;	/* don't merge multiple soundfiles into one */

CLE_TOKEN sampfmtTable[] = {
    { "c?har", SFMT_CHAR }, 
    { "o?ffsetchar", SFMT_CHAR_OFFS }, 
    { "a?law", SFMT_ALAW }, 
    { "u?law|mu?law", SFMT_ULAW }, 
    { "s?hort|i16", SFMT_SHORT }, 
    { "m?ed|i24", SFMT_24BL }, 
    { "l?ong|i32", SFMT_LONG }, 
    { "f?loat|f32", SFMT_FLOAT }, 
    { "d?ouble|f64", SFMT_DOUBLE },
    { NULL, 0}
};

CLE_TOKEN chanmodeTable[] = {
    { "m?ix|mo?no", SCMD_MONO }, 
    { "a|A|0|l?eft|L?EFT", SCMD_CHAN0 }, 
    { "b|B|1|r?ight|R?IGHT", SCMD_CHAN1 }, 
    { "c|C|2", SCMD_CHAN2 }, 
    { "d|D|3", SCMD_CHAN3 }, 
    { "e|E|4", SCMD_CHAN4 }, 
    { "f|F|5", SCMD_CHAN5 }, 
    { "g|G|6", SCMD_CHAN6 }, 
    { "h|H|7", SCMD_CHAN7 }, 
    { NULL, 0}
};

CLE_ENTRY clargs[] = {
    { "-o?utputfile", CLE_T_STRING, "Output file (NOTE NEW SYNTAX!)", 
      "-", &outputfilename, NULL, NULL },

    { "-S|-ip?format", CLE_T_STRING, "input sound file format (sndf token)", 
      NULL, &sfifmtname, NULL, NULL}, 
    { "-T|-op?format", CLE_T_STRING, "output sound file format (sndf token)", 
      NULL, &sfofmtname, NULL, NULL}, 

    { "-k|-sk?ip", CLE_T_TIME, "skip this many seconds of initial sound", 
      "0.0", &skiptime, NULL, NULL },
    { "-d?uration", CLE_T_TIME, "force output duration to this many secs", 
      "0.0", &durtime, NULL, NULL }, 
    { "-e?ndtime", CLE_T_TIME, "stop copying at this point in file (secs)", 
      "0.0", &endtime, NULL, NULL }, 
    { "-K|-skipf?rames", CLE_T_INT, "skip this many sample frames initially", 
      "0", &skipframes, NULL, NULL },  
    { "-D|-durf?rames", CLE_T_INT, "force duration to this many frames", 
      "0", &durframes, NULL, NULL }, 
    { "-E|-endf?rame", CLE_T_INT, "absolute endpoint in samples", 
      "0", &endframe, NULL, NULL }, 
    { "-c?hans|-chann?els", CLE_T_INT, "force output to this many channels", 
      "0", &chans, NULL, NULL }, 
    { "-chanm?ode|-cm?ode", CLE_T_TOKEN, "if converting to/from mono, set mode", 
      "mono", &chanmode, NULL, chanmodeTable }, 
    { "-f?ormat|-fmt", CLE_T_TOKEN, "output sample format", 
      NULL, &sampfmt, NULL, sampfmtTable }, 
    { "-r?ate|-sr?ate|-s?amplerate", CLE_T_FLOAT, "force output sample rate", 
      NULL, &srate, NULL, NULL }, 
    { "-g?ain", CLE_T_FLOAT, "apply gain factor to data", 
      "1.0", &gain, NULL, NULL },
    { "-n?omerge", CLE_T_BOOL, "don't merge all sounds into a single unsegmented output sound, but write segment separators inbetween", 
      "FALSE", &nomerge, NULL, NULL }, 
    { "-q?uery", CLE_T_BOOL, "just query soundfile(s)", 
      "FALSE", &query, NULL, NULL }, 
    { "-v?erbose", CLE_T_BOOL, "verbose information reporting", 
      "FALSE", &verb, NULL, NULL }, 
    { "-mu?lti", CLE_T_BOOL, "attempt to read multiple soundfiles from stream", 
      "FALSE", &domulti, NULL, NULL }, 
    { "-h?elp", CLE_T_BOOL, "print help message & exit", "FALSE", &help, NULL, NULL },


    { "[inputfiles ...]", CLE_T_USAGE, NULL, 
     "-", NULL, NULL}, 
    { NULL, CLE_T_END, NULL, NULL, NULL, NULL, NULL }
};

int main(int argc, char *argv[])
{
    int 	i = 0;
    int		err;
    int 	j;
    char *programName = argv[0];
    SOUND *isnd = NULL, *osnd = NULL;
    SOUND *isndtplt;	/* to hold default sfifmt */
    int ipchans, ipformat, ipframes, ipsffmt;
    float ipsrate, lastipsrate = 0;
    char **pipfname;
    char *ipfname = NULL;
    float *buf = NULL;
    int bufsize = 32768;
    int bufbytes;
    int remains, doneframes, totdoneframes, got, toget;
    int firsttime;
    int filestodo, filesdone;
    int lenset;
    int domerge = 1;

    /*  Handle the args */
    cleSetDefaults(clargs);
    err = cleExtractArgs(clargs, &argc, argv, 0);
    if(err) {
	cleUsage(clargs, programName);
	exit(-1);
    }
    if(help) {
	cleUsage(clargs, programName);
	exit(0);
    }

    /* post-process */
    if (argc > 1) {
        pipfname = &(argv[1]);
	filestodo = argc - 1;
    } else {
	/* No input files specified - use stdin */
	ipfname = "-";
	pipfname = &ipfname;
	filestodo = 1;
    }
    filesdone = 0;
    totdoneframes = 0;

    if (nomerge) {
	domerge = 0;
    }

    /* Input soundfile fmt name; accept "unk?nown" meaning to guess */
    if (sfifmtname && \
	strncmp(sfifmtname, "unknown", strlen(sfifmtname)) != 0) {
        if (!SFSetDefaultFormatByName(sfifmtname)) {
	    fprintf(stderr, "sndffmt: '%s' is not recognized.  Try:\n", 
		    sfifmtname);
	    SFListFormats(stderr);
	    fprintf(stderr, "\n");
	    cleUsage(clargs, programName);
	    exit(-1);
	}
    } else {
        /* no input type specified, so set to guess */
	SFSetDefaultFormat(SFF_UNKNOWN);
    }	

    /* preset the input format as requested (alternative to SF* stuff) */
    isndtplt = sndNew(NULL);
    if (sfifmtname) sndSetSFformatByName(isndtplt, sfifmtname);

    /* cycle through all the input file names */
    while ( filesdone < filestodo ) {

	ipfname = *pipfname++;

	isnd = sndNew(isndtplt);
	if( (isnd = sndOpen(isnd, ipfname, "r")) == NULL)  {
	    /* could not open as sound file */
	    fprintf(stderr, "%s: couldn't open '%s' for input\n", 
		    programName, ipfname);
	    exit(-1);

	} else {

	    /* now we have ipsrate, can maybe convert times to frames */
	    ipsrate = sndGetSrate(isnd);
	    if (lastipsrate == 0) {
		lastipsrate = ipsrate;
	    } else if (ipsrate != lastipsrate) {
		/* different sample rate of sucessive sounds. 
		   Not a problem unless we're sticking them together */
		if (domerge) {
		    fprintf(stderr, "%s: warn: sample rate of input changed from %.0f to %.0f\n", programName, lastipsrate, ipsrate);
		}
	    }
	    if (skipframes == 0 && skiptime > 0.0) {
		skipframes = (int)(rint(ipsrate * skiptime));
	    }
	    /* durframes and durtime take precedence over endtime */
	    if (durtime == 0.0 && durframes == 0 \
		&& endtime > 0.0 && endtime > skiptime) {
		durtime = endtime - skiptime;
	    }
	    if (durframes == 0 && durtime > 0.0) {
		durframes = (int)(rint(ipsrate * durtime));
	    }
	    /* endframe is respected only if endtime and dur* are not set */
	    if (durframes == 0 && endframe > 0 && endframe > skipframes) {
		durframes = endframe - skipframes;
	    }
	    ipchans   = sndGetChans(isnd);
	    if (chans > 0)  {
		sndSetUchans(isnd, chans);   /* force e.g. to mono */
		ipchans = chans;
	    }
	    /* Set channel mode too (or only for mapping to 1 ?) */
	    sndSetChanmode(isnd, chanmode);

	    bufbytes = sizeof(float)*ipchans*bufsize;
	    if( (buf = malloc(bufbytes)) == NULL) {
		fprintf(stderr, "error allocating buffer\n");
		exit (-1);
	    }
	    memset(buf, 0, bufbytes);

	    if (!query) {
		if (osnd == NULL) {
		    /* create output sound */
		    osnd = sndNew(isnd);
		    if (srate > 0) {	sndSetSrate(osnd, srate);	}
		    if (chans > 0) {	sndSetChans(osnd, chans);	}
		    if (sampfmt > 0) {	sndSetFormat(osnd, sampfmt);	}
		}
		if (osnd->isOpen && !domerge) {
		    /* checkpointing; do this *before* setting the length */
		    sndNext(osnd);	
		}
		lenset = 0;
		if (durframes > 0) {    
		    sndSetFrames(osnd, durframes);  
		    lenset = 1;
		} else {
		    int iframes = sndGetFrames(isnd);
		    if (iframes != SF_UNK_LEN && (filestodo == 1 || !domerge)) {
			sndSetFrames(osnd, iframes - skipframes);
			lenset = 1;
		    }
		}
		if (!osnd->isOpen) {
		    ipsffmt = SFGetFormat(isnd->file, NULL);
		    if(sfofmtname) {
  		        if (!sndSetSFformatByName(osnd, sfofmtname)) {
			    fprintf(stderr, "sndfofmt: '%s' is not recognized.  Try:\n", 
				    sfofmtname);
			    SFListFormats(stderr);
			    fprintf(stderr, "\n");
			    cleUsage(clargs, programName);
			    exit(-1);
			}
		    } else {
			sndSetSFformat(osnd, ipsffmt);
		    }
		    if (sndOpen(osnd, outputfilename, "w") == NULL) {
			fprintf(stderr, "%s: couldn't write to '%s'\n", 
				programName, outputfilename);
			exit(-1);
		    }
		}
	    }

	    /* sndNext is a flag that checks to see if there are 
	       more soundfiles on this single pipe */
	    firsttime = 1;
	    while (err == 0 && (firsttime || (domulti && sndNext(isnd)))) {
		firsttime = 0;

		if(verb) {
		    sndPrint(isnd, stderr, "");
		}

		ipchans   = sndGetChans(isnd);
		ipformat  = sndGetFormat(isnd);
		ipframes  = sndGetFrames(isnd);
		ipsrate   = sndGetSrate(isnd);

		if (!query || domulti) {

		    if(skipframes) {
			sndFrameSeek(isnd, skipframes, SEEK_SET);
		    }
		    if(durframes > 0) {
			if (ipframes > durframes) {
			    ipframes = durframes;
			} else if (ipframes < 0) {
			    ipframes = durframes;
			}
		    }
		    if (srate > 0)	
			ipsrate = srate;	/* override file's sr */
		    if (chans > 0)  {
			sndSetUchans(isnd, chans);   /* force e.g. to mono */
			ipchans = chans;
		    }

		    doneframes = 0;
		    /* actually copy the samples from file to output */
		    if(ipframes > 0)
			remains = ipframes;
		    else
			remains = -1;
		    do {
			if(remains > 0)
			    toget = MIN(bufsize, remains);
			else
			    toget = bufsize;
			got = sndReadFrames(isnd, buf, toget);

			if (gain != 1.0) {
			    for (j = 0; j < got * ipchans; ++j) {
				buf[j] *= gain;
			    }
			}

			if(osnd) sndWriteFrames(osnd, buf, got);

			doneframes += got;
			if(remains > 0)
			    remains -= got;
		    } while(remains != 0 && got > 0);

		    /* pad out end with zeros if req'd duration longer */
		    if (durframes > 0 && doneframes < durframes) {
			for (j = 0; j < bufsize; ++j) {
			    buf[j] = 0;
			}
			while (doneframes < durframes) {
			    remains = durframes - doneframes;
			    got = MIN(bufsize, remains);
			    if(osnd) sndWriteFrames(osnd, buf, got);
			    doneframes += got;
			}
		    }

		    if(verb) {
			fprintf(stderr, "(%d frames)\n", doneframes);
		    }
		    totdoneframes += doneframes;
		}
	    }
	}
	if (osnd && !lenset && !domerge) {
	    /* attempt to write length retrospectively */
	    sndSetFrames(osnd, doneframes);
	    lenset = 1;
	}
	++filesdone;
    }
    if (osnd && !lenset) {
	/* write length only at the very end */
	sndSetFrames(osnd, totdoneframes);
    }
    if(buf)	free(buf);
    if(osnd)	sndClose(osnd);
    if(isnd)	sndClose(isnd);

    exit(0);
}
